### PR TITLE
LIP-7: a composite beacon report receiver

### PR DESCRIPTION
**Pull request by TheDZhon**
Propose a composite beacon report receiver to allow multiple callbacks
invocation per single lido oracle beacon report run.

### CODE REVIEW

File: LIPS/lip-composite_beacon_report_receiver.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+# Introduce a composite oracle beacon report receiver
+
+## Simple summary
+
+To have a possibility of using two or even more Lido Oracle beacon report intercepting callbacks, we introduce a composite oracle beacon report receiver. The last one implements the `IBeaconReceiver` interface and internally holds an array of nested `IBeaconReceiver` instances and supplies iterative execution by calling `processLidoOracleReport` in a loop.
+
+## Abstract
+
+The `OrderedCallbacksArray` contract implements the aforementioned `IOrderedCallbacksArray` interface adding the access modifiers to allow storage changing calls (add/insert/remove) only originated by the `Voting` contract (having `require(msg.sender == Voting.address)` ). Another one contract (`CompositePostRebaseBeaconReceiver`) inherits `OrderedCallbacksArray` and implemens `IBeaconReportReceiver` interface while allowing to call `processLidoOracleReport` function only originated by the `LidoOracle` contract (having `require(msg.sender == LidoOracle.address)`).
```
Comment (folkyatina) [inline LIPS/lip-composite_beacon_report_receiver.md]: s/Another one contract/Another contract

Comment (folkyatina) [reply LIPS/lip-composite_beacon_report_receiver.md]: implemen**t**s

Comment (folkyatina) [reply LIPS/lip-composite_beacon_report_receiver.md]: The last sentence can look better being split into smaller, simpler ones.

File: LIPS/lip-composite_beacon_report_receiver.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+In the end, the `CompositePostRebaseBeaconReceiver` contract is a top-level entity implementing the current proposal.
+
+## Motivation
+
+Currently, the `LidoOracle` [contract](https://docs.lido.fi/contracts/lido-oracle) provides only one slot for a quorum report intercepting callback (see [LidoOracle.setBeaconReportReceiver](https://docs.lido.fi/contracts/lido-oracle#setbeaconreportreceiver)). We plan to occupy this slot soon with a newly developed contract implementing the [LIP-6 coverage application mechanism](https://research.lido.fi/t/lip-6-in-protocol-coverage-proposal/1468). In contrast, we have a technical vision of the future cross-chain/L2 [upgrades](https://hackmd.io/f3416OoaS2e1l2xu_bX6SA), which may require append additional callbacks propagating oracle beacon reports.
```
Comment (folkyatina) [inline LIPS/lip-composite_beacon_report_receiver.md]: require **to** append ?

File: LIPS/lip-composite_beacon_report_receiver.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+```
+Removes element with a given by the `_atIndex` param position from the callbacks array.
+Elements after the `_atIndex` position are shifted one position left to preserve an existing invocation order.
+* Reverts if `_atIndex` is equal to or greater than the length of the callbacks array.
+* Reverts if `msg.sender` is not equal to the stored `voting` address.
+* Emits the `CallbackRemoved` event.
+See: `callbacksLength`.
+
+```solidity
+function callbacks(uint256 _atIndex) external view returns (address);
```
Comment (arwer13) [inline LIPS/lip-composite_beacon_report_receiver.md]: Imho, `getCallback` or `callbackAt` would be a clearer name

File: LIPS/lip-composite_beacon_report_receiver.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+Init contract and store the provided `_voting` address to arm the `onlyVoting` modifier.
+See: `onlyVoting`.
+
+```solidity
+function addCallback(address _callback) external override onlyVoting;
+```
+Adds the provided `_callback` element at the end of the callbacks array, after its current last element.
+* Reverts if `_callback` address is zero.
+* Reverts if `msg.sender` is not equal to the stored `voting` address.
+* Reverts if the new length of the callbacks array becomes greater than the `MAX_CALLBACKS_CNT` contract-wide constant.
```
Comment (arwer13) [inline LIPS/lip-composite_beacon_report_receiver.md]: Maybe make `uint256 public constant MAX_CALLBACKS_COUNT = 16;` an immutable? To be able to parametrize it on contract creation and to reduce storage usage a bit.

https://github.com/lidofinance/lido-dao/blob/e935ce24cf373f79bdf9d366c714a9cb7ca7e08b/contracts/0.8.9/OrderedCallbacksArray.sol#L17

File: LIPS/lip-composite_beacon_report_receiver.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+## Security considerations
+
+#### Upgradability
+The proposed contracts are non-upgradable for the sake of simplicity. In case of emergency, we can disable them entirely by calling `LidoOracle.setBeaconReportReceiver(address(0))` and redeploy newly developed versions later from scratch without incurring additional proxy-initialization logic and state transitions.
+
+#### Access control (permissions model)
+There are two permissioned addresses introduced explicitly: `Voting` and `LidoOracle`. Only `Voting` can add/insert/remove callbacks into the `CompositePostReabaseBeaconReceiver` to make it feasible only by the Lido DAO direct vote-backed will. Only `LidoOracle` is allowed to call `processLidoOracleReport` to prevent propagating various fake oracle reports by the added callbacks or execution in an arbitrary time moment.
+
+#### Misbehaving callback
```
Comment (arwer13) [inline LIPS/lip-composite_beacon_report_receiver.md]: I propose to note here that if a callback throws an exception all the subsequent callbacks get skipped.

File: LIPS/lip-composite_beacon_report_receiver.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+```solidity
+constructor(address _voting)
+```
+Init contract and store the provided `_voting` address to arm the `onlyVoting` modifier.
+See: `onlyVoting`.
+
+```solidity
+function addCallback(address _callback) external override onlyVoting;
+```
+Adds the provided `_callback` element at the end of the callbacks array, after its current last element.
```
Comment (folkyatina) [inline LIPS/lip-composite_beacon_report_receiver.md]: Probably, the last part "after its current last element" is excessive.

**Review by folkyatina — COMMENTED** [review]
Added some remarks on grammar and style, that could be safely ignored in some parts. 

Beside of that, everything is clear. Maybe some comments on chosen data structure can be useful, but on the other hand it seems obvious and can be omitted safely.

File: LIPS/lip-composite_beacon_report_receiver.md (hunk 3 above)
Comment (TheDZhon) [reply LIPS/lip-composite_beacon_report_receiver.md]: Yeah, you are completely right.
But this one is autogenerated getter for the public `callbacks` storage var.
So I just decided to make it consistent from this point of view.

File: LIPS/lip-composite_beacon_report_receiver.md (hunk 4 above)
Comment (TheDZhon) [reply LIPS/lip-composite_beacon_report_receiver.md]: Hmm, as I know, both `immutable` and `constant` are occupy zero storage slot space.

I have found also this answer https://ethereum.stackexchange.com/a/113885 which highlights the substitutions difference.

Nevertheless, we set it to `16` just as a sanity check.

File: LIPS/lip-composite_beacon_report_receiver.md (hunk 5 above)
Comment (TheDZhon) [reply LIPS/lip-composite_beacon_report_receiver.md]: Good point, I'll add it.

File: LIPS/lip-composite_beacon_report_receiver.md (hunk 6 above)
Comment (TheDZhon) [reply LIPS/lip-composite_beacon_report_receiver.md]: Yeah, you are right :+1:

File: LIPS/lip-composite_beacon_report_receiver.md (hunk 2 above)
Comment (TheDZhon) [reply LIPS/lip-composite_beacon_report_receiver.md]: paraphrased it

File: LIPS/lip-composite_beacon_report_receiver.md (hunk 1 above)
Comment (TheDZhon) [reply LIPS/lip-composite_beacon_report_receiver.md]: Yeah, thx! Fixed.

File: LIPS/lip-composite_beacon_report_receiver.md (hunk 4 above)
Comment (arwer13) [reply LIPS/lip-composite_beacon_report_receiver.md]: Indeed!

**Pull request merged**